modules/syslogformat/filterx-func-format-syslog.c

```diff
@@ -40,11 +49,238 @@ typedef struct FilterXFunctionFormatSyslog5424_
   FilterXExpr *message_expr;
 } FilterXFunctionFormatSyslog5424;
 
+static inline FilterXObject *
+_eval_optional_or_fallible_expr(FilterXExpr *expr)
+{
+  if (!expr)
+    return NULL;
+
+  FilterXObject *obj = filterx_expr_eval(expr);
+  if (!obj)
+    filterx_eval_clear_errors();
+
+  return obj;
+}
+
+static inline const gchar *
+_cast_to_string(FilterXObject *obj, gsize *len)
+{
+  if (!obj)
+    {
+      *len = 0;
+      return "";
+    }
+
+  const gchar *str;
+  if (filterx_object_extract_string_ref(obj, &str, len))
+    return str;
+
+  GString *buf = scratch_buffers_alloc();
+  if (filterx_object_str(obj, buf))
+    {
+      *len = buf->len;
+      return buf->str;
+    }
+
+  filterx_eval_push_error("Cannot convert object to string", NULL, obj);
+  return NULL;
+}
+
+static inline gint64
+_cast_to_int(FilterXObject *obj)
+{
+  if (!obj)
+    return -1;
+
+  gint64 value;
+  if (filterx_object_extract_integer(obj, &value))
+    return value;
+
+  return -1;
+}
+
+static inline FilterXObject *
+_get_message(FilterXFunctionFormatSyslog5424 *self, const gchar **message, gsize *message_len)
+{
+  FilterXObject *message_obj = filterx_expr_eval(self->message_expr);
+  if (!message_obj)
+    return NULL;
+  *message = _cast_to_string(message_obj, message_len);
+  return message_obj;
+}
+
+static inline FilterXObject *
+_get_pri(FilterXFunctionFormatSyslog5424 *self, LogMessage *logmsg, gint64 *pri)
+{
+  FilterXObject *pri_obj = _eval_optional_or_fallible_expr(self->pri_expr);
+  *pri = _cast_to_int(pri_obj);
+  if (*pri == -1)
+    *pri = logmsg->pri;
+  return pri_obj;
+}
+
+static inline FilterXObject *
+_get_timestamp(FilterXFunctionFormatSyslog5424 *self, LogMessage *logmsg, WallClockTime *timestamp)
+{
+  UnixTime ut;
+  FilterXObject *timestamp_obj = _eval_optional_or_fallible_expr(self->timestamp_expr);
+  if (!timestamp_obj || !filterx_object_extract_datetime(timestamp_obj, &ut))
+    ut = logmsg->timestamps[LM_TS_STAMP];
+  convert_unix_time_to_wall_clock_time(&ut, timestamp);
+  return timestamp_obj;
+}
+
+static inline FilterXObject *
+_get_host(FilterXFunctionFormatSyslog5424 *self, const gchar **host, gsize *host_len)
+{
+  FilterXObject *host_obj = _eval_optional_or_fallible_expr(self->host_expr);
+  *host = _cast_to_string(host_obj, host_len);
+  if (*host_len == 0)
+    {
+      *host = get_local_hostname_fqdn();
```

Comment:
I think we should consider using $HOST in this case. All the other defaults extract information from the
log message.

The local hostname is almost always incorrect, I think.